File: src/providers/deepseek-client.ts
import type { ChatMessage, ToolDef, CompletionResult } from "../types.ts";

export interface StreamCallbacks {
  onDelta: (text: string) => void;
  onToolCallDelta: (delta: { index: number; id?: string; name?: string; argsFragment?: string }) => void;
}

export async function complete(
  model: string,
  messages: ChatMessage[],
  tools: ToolDef[] = [],
  apiKey: string = process.env.DEEPSEEK_API_KEY ?? "",
  baseUrl: string = process.env.DEEPSEEK_BASE_URL ?? "https://api.deepseek.com",
  fetchImpl: typeof fetch = fetch
): Promise<CompletionResult> {
  if (!apiKey) {
    throw new Error("DEEPSEEK_API_KEY not set");
  }

  const body: Record<string, unknown> = { model, messages };
  if (tools.length > 0) {
    body.tools = tools;
  }

  const response = await fetchImpl(`${baseUrl}/v1/chat/completions`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify(body),
  });

  if (response.status !== 200) {
    throw new Error(`DeepSeek API returned ${response.status}: ${await response.text()}`);
  }

  const parsed = (await response.json()) as {
    choices: { message: { content: string | null; tool_calls?: CompletionResult["toolCalls"] } }[];
  };
  const message = parsed.choices[0].message;
  return { content: message.content ?? "", toolCalls: message.tool_calls ?? [] };
}

export async function completeStream(
  model: string,
  messages: ChatMessage[],
  tools: ToolDef[] = [],
  callbacks: StreamCallbacks,
  signal?: AbortSignal,
  apiKey: string = process.env.DEEPSEEK_API_KEY ?? "",
  baseUrl: string = process.env.DEEPSEEK_BASE_URL ?? "https://api.deepseek.com",
  fetchImpl: typeof fetch = fetch
): Promise<CompletionResult> {
  if (!apiKey) {
    throw new Error("DEEPSEEK_API_KEY not set");
  }

  const body: Record<string, unknown> = { model, messages, stream: true };
  if (tools.length > 0) {
    body.tools = tools;
  }

  const response = await fetchImpl(`${baseUrl}/v1/chat/completions`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify(body),
    signal,
  });

  if (response.status !== 200) {
    throw new Error(`DeepSeek API returned ${response.status}: ${await response.text()}`);
  }
  if (!response.body) {
    throw new Error("DeepSeek API returned a streaming response with no body");
  }

  let content = "";
  const toolCallsByIndex = new Map<number, { id: string; name: string; args: string }>();
  const decoder = new TextDecoder();
  let buffer = "";

  for await (const chunk of response.body) {
    buffer += decoder.decode(chunk as Uint8Array, { stream: true });
    let boundary = buffer.indexOf("\n\n");
    while (boundary !== -1) {
      const rawEvent = buffer.slice(0, boundary);
      buffer = buffer.slice(boundary + 2);
      boundary = buffer.indexOf("\n\n");

      const line = rawEvent.trim();
      if (!line.startsWith("data:")) continue;
      const payload = line.slice("data:".length).trim();
      if (payload === "[DONE]") continue;

      const parsed = JSON.parse(payload) as {
        choices: { delta: { content?: string; tool_calls?: { index: number; id?: string; function?: { name?: string; arguments?: string } }[] } }[];
      };
      const delta = parsed.choices[0]?.delta;
      if (!delta) continue;

      if (delta.content) {
        content += delta.content;
        callbacks.onDelta(delta.content);
      }

      if (delta.tool_calls) {
        for (const tc of delta.tool_calls) {
          const existing = toolCallsByIndex.get(tc.index) ?? { id: tc.id ?? "", name: tc.function?.name ?? "", args: "" };
          if (tc.id) existing.id = tc.id;
          if (tc.function?.name) existing.name = tc.function.name;
          if (tc.function?.arguments) existing.args += tc.function.arguments;
          toolCallsByIndex.set(tc.index, existing);
          callbacks.onToolCallDelta({ index: tc.index, id: tc.id, name: tc.function?.name, argsFragment: tc.function?.arguments });
        }
      }
    }
  }

  const toolCalls = [...toolCallsByIndex.entries()]
    .sort(([a], [b]) => a - b)
    .map(([, tc]) => ({ id: tc.id, type: "function" as const, function: { name: tc.name, arguments: tc.args } }));

  return { content, toolCalls };
}

File: MADE/core/experiment/deepseek_client.py
import os
import time
from dataclasses import dataclass

import httpx


class DeepSeekError(RuntimeError):
    pass


@dataclass
class CompletionResult:
    text: str
    cost_usd: float
    latency_ms: float


_PRICING_USD_PER_1K_TOKENS = {
    "deepseek-v4-flash": 0.001,
    "deepseek-v4-pro": 0.01,
}

_MAX_RETRIES = 3
_RETRY_BACKOFF_SECONDS = 1.0


def _base_url() -> str:
    return os.environ.get("DEEPSEEK_BASE_URL", "https://api.deepseek.com")


def _api_key() -> str:
    key = os.environ.get("DEEPSEEK_API_KEY")
    if not key:
        raise DeepSeekError("DEEPSEEK_API_KEY not set")
    return key


def complete(model: str, prompt: str) -> CompletionResult:
    if model not in _PRICING_USD_PER_1K_TOKENS:
        raise DeepSeekError(f"unknown pricing for model {model!r}")
    api_key = _api_key()

    last_error: Exception | None = None
    for attempt in range(_MAX_RETRIES):
        start = time.monotonic()
        try:
            response = httpx.post(
                f"{_base_url()}/v1/chat/completions",
                headers={"Authorization": f"Bearer {api_key}"},
                json={"model": model, "messages": [{"role": "user", "content": prompt}]},
                timeout=60,
            )
            latency_ms = (time.monotonic() - start) * 1000
            if response.status_code != 200:
                raise DeepSeekError(f"DeepSeek API returned {response.status_code}: {response.text}")
            body = response.json()
            text = body["choices"][0]["message"]["content"]
            total_tokens = body["usage"]["total_tokens"]
            cost_usd = (total_tokens / 1000) * _PRICING_USD_PER_1K_TOKENS[model]
            return CompletionResult(text=text, cost_usd=cost_usd, latency_ms=latency_ms)
        except (httpx.TimeoutException, httpx.NetworkError, DeepSeekError) as exc:
            last_error = exc
            if attempt < _MAX_RETRIES - 1:
                time.sleep(_RETRY_BACKOFF_SECONDS)

    raise DeepSeekError(f"DeepSeek API call failed after {_MAX_RETRIES} attempts: {last_error}") from last_error
